sips/sip-clarity4/sip-clarity4.md

```diff
@@ -0,0 +1,404 @@
+    before executing the body expressions. The final body expression cannot
+    return a `response` value in order to avoid returning a nested `response`
+    value from the `restrict-assets?` (nested responses are error-prone).
+    Returns:
+
+    - `(ok x)` if the outflows are within the allowances, where `x` is the
+      result of the `body` expression and has type `A`.
+    - `(err index)` if an allowance was violated, where `index` is the 0-based
+      index of the first violated allowance in the list of granted allowances,
+      or -1 if an asset with no allowance caused the violation.
+
+  - **Example**:
+    ```clarity
+    (define-public (foo)
+      (as-contract? ()
+        (try! (stx-transfer? u1000000 tx-sender recipient))
+      )
+    ) ;; Returns (err -1)
+    (define-public (bar)
+      (as-contract? ((with-stx u1000000))
+        (try! (stx-transfer? u1000000 tx-sender recipient))
+      )
+    ) ;; Returns (ok true)
+    ```
+
+- `current-contract`
+  - **Description**: Returns the principal of the current contract.
+  - **Output**: `principal`
+  - **Example**:
+    ```clarity
+    (stx-transfer? u1000000 tx-sender current-contract)
+    ```
+
+## Conversion to `string-ascii`: `to-ascii?`
+
+Originally proposed here: https://github.com/clarity-lang/reference/issues/82
+
+`to-ascii?` is a new Clarity function that converts simple values into their
+`string-ascii` representations.
+
+- **Input**: `int` | `uint` | `bool` | `principal` | `(buff 524284)` |
+  `(string-utf8 262144)`
+- **Output**: `(response (string-ascii 1048571) uint)`
+- **Signature**: `(to-ascii? value)`
+- **Description**: Returns the `string-ascii` representation of the input value
+  in an `ok` response on success. The only error condition is if the input type
+  is `string-utf8` and the value contains non-ASCII characters, in which case,
+  `(err u1)` is returned.
+- **Example**:
+  ```clarity
+  (to-ascii? true) ;; Returns (ok "true")
+  (to-ascii? 42) ;; Returns (ok "42")
+  (to-ascii? 'SP2QEZ06AGJ3RKJPBV14SY1V5BBFNAW33D96YPGZF) ;; Returns (ok "SP2QEZ06AGJ3RKJPBV14SY1V5BBFNAW33D96YPGZF")
+  (to-ascii? 0x12345678) ;; Returns (ok "0x12345678")
+  ```
+
+## Timestamp for current block: `block-time`
+
+`block-time` is a new Clarity keyword that returns the timestamp of the current
+block in seconds since the Unix epoch. This is the same timestamp that is in the
+block header, which is verified by the signers to be:
+
+- Greater than the timestamp of its parent block
+- At most 15 seconds into the future
+
+This same timestamp can also be retrieved for previous blocks using
+`(get-stacks-block-info? time height)`, which exists since Clarity 3, but cannot
+be used for the current block.
+
+- **Output**: `uint`
+- **Example**:
+  ```clarity
+  (if (> block-time 1755820800)
+    (print "after 2025-07-22")
+    (print "before 2025-07-22"))
+  ```
+
+# Related Work
+
+Not applicable.
+
+# Backwards Compatibility
```

Comment:
I'm not sure if it is worth including, but `current-contract` will become a keyword in Clarity4. Any contract code predating it needs to be slightly refactored if it is to be redeployed as Clarity4. Luckily, it does appear that `current-contract` is [not used as a variable or constant name often](https://github.com/search?q=repo%3Aboomcrypto%2Fclarity-deployed-contracts%20current-contract&type=code).

(Disregarding @LNow's black magic that allows you to override keywords in certain situations.)